nixos/modules/services/x11/redshift.nix

```diff
@@ -106,31 +150,51 @@ in {
     # needed so that .desktop files are installed, which geoclue cares about
     environment.systemPackages = [ cfg.package ];
 
-    services.geoclue2.appConfig.redshift = {
+    services.geoclue2.appConfig.redshift = mkIf isGeoclue2 {
       isAllowed = true;
       isSystem = true;
     };
 
     systemd.user.services.redshift =
     let
-      providerString = if lcfg.provider == "manual"
-        then "${toString lcfg.latitude}:${toString lcfg.longitude}"
-        else lcfg.provider;
+      mainSection = if strings.hasInfix "gammastep" cfg.executable
+        # https://gitlab.com/chinstrap/gammastep/-/commit/1608ed61154cc652b087e85c4ce6125643e76e2f
+        then "general"
+        else "redshift";
+
+      locationConfig = if isManual
+        then {
+          manual = {
+            lat = lcfg.latitude;
+            lon = lcfg.longitude;
+          };
+        } else {};
+
+      mainConfig = {
+        ${mainSection} = {
+          temp-day = cfg.temperature.day;
+          temp-night = cfg.temperature.night;
+          fade = if cfg.fade then "1" else "0";
+          location-provider = lcfg.provider;
+          brightness-day = cfg.brightness.day;
+          brightness-night = cfg.brightness.night;
+          gamma-day = cfg.gamma.day;
+          gamma-night = cfg.gamma.night;
+        };
+      };
+
+      fullConfig = mainConfig // locationConfig // cfg.extraConfig;
+      iniText = generators.toINI {} fullConfig;
+      configFile = pkgs.writeText "redshift.conf" iniText;
     in
     {
       description = "Redshift colour temperature adjuster";
       wantedBy = [ "graphical-session.target" ];
       partOf = [ "graphical-session.target" ];
       serviceConfig = {
-        ExecStart = ''
-          ${cfg.package}${cfg.executable} \
-            -l ${providerString} \
-            -t ${toString cfg.temperature.day}:${toString cfg.temperature.night} \
-            -b ${toString cfg.brightness.day}:${toString cfg.brightness.night} \
-            ${lib.strings.concatStringsSep " " cfg.extraOptions}
-        '';
+        ExecStart = "${cfg.package}${cfg.executable} -c ${configFile}";
         RestartSec = 3;
-        Restart = "always";
+        Restart = "on-failure";
```

Comment:
This avoids the service to keep restarting in case of an error in config (that may happen for example when setting a wrong `extraConfig` value).